log/util.go

```diff
@@ -3,27 +3,52 @@ package log
 import (
 	"context"
 
-	"github.com/anz-bank/pkg/log/loggers"
-	"github.com/marcelocantos/frozen"
+	"github.com/arr-ai/frozen"
 )
 
-func getCopiedLogger(ctx context.Context) loggers.Logger {
-	logger, exists := ctx.Value(loggerKey).(loggers.Logger)
+type loggerContextKey struct{}
+type fieldsContextKey struct{}
+type suppress struct{}
+type ctxRef struct{ ctxKey interface{} }
+
+func getCopiedLogger(ctx context.Context) Logger {
+	logger, exists := ctx.Value(loggerContextKey{}).(Logger)
 	if !exists {
 		panic("Logger does not exist in context")
 	}
-	return logger.Copy()
+	return logger.(internalLoggerOps).Copy()
 }
 
-func fromFields(fields []Fields) frozen.Map {
-	// set capacity to 0 because length of fields is undetermined
-	frozenMapBuilder := frozen.NewMapBuilder(0)
-
-	for _, f := range fields {
-		for _, kv := range f.GetKeyValues() {
-			frozenMapBuilder.Put(kv.Key, kv.Value)
+func (f Fields) resolveFields(ctx context.Context) frozen.Map {
+	fields := f.m
+	var toSuppress frozen.SetBuilder
+	for i := fields.Range(); i.Next(); {
+		switch k := i.Value().(type) {
+		case ctxRef:
+			if val := ctx.Value(k.ctxKey); val != nil {
+				fields = fields.With(i.Key(), val)
+			}
+		case func(context.Context) interface{}:
+			fields = fields.With(i.Key(), k(ctx))
+		case suppress:
+			toSuppress.Add(i.Key())
 		}
 	}
+	return fields.Without(toSuppress.Finish())
+}
+
+func (f Fields) merge(fs ...Fields) Fields {
```

Comment:
Does this need to exist? It's identical to `Chain`.

Reply:
yep just realized that, moved `merge` implementation to `Chain`